client/client.go

```diff
@@ -1255,32 +1265,68 @@ func (r *repository) pubKeyListForRotation(role data.RoleName, serverManaged boo
 	}
 
 	// If no new keys are passed in, we generate one
-	if len(newKeys) == 0 {
+	if len(newKeys) == 0 && len(pubKeys) == 0 {
 		pubKeyList = make(data.KeyList, 0, 1)
 		pubKey, err = r.GetCryptoService().Create(role, r.gun, data.ECDSAKey)
 		pubKeyList = append(pubKeyList, pubKey)
 	}
+
+	// If a list of keys to rotate to are provided, we add those
+	if len(newKeys) > 0 || len(pubKeys) > 0 {
+		pubKeyList, err = r.pubKeyListFromPrivAndPubKeys(newKeys, pubKeys)
+	}
 	if err != nil {
 		return nil, fmt.Errorf("unable to generate key: %s", err)
 	}
 
 	// If a list of keys to rotate to are provided, we add those
-	if len(newKeys) > 0 {
-		pubKeyList = make(data.KeyList, 0, len(newKeys))
-		for _, keyID := range newKeys {
-			pubKey = r.GetCryptoService().GetKey(keyID)
+	if pubKeyList, err = r.pubKeysToCerts(role, pubKeyList); err != nil {
+		return nil, err
+	}
+
+	return pubKeyList, nil
+}
+
+// pubKeyListFromPrivAndPubKeys returns a list of consolidated public keys obtained by
+// merging the pubKeys with the public keys of the private keys
+func (r *repository) pubKeyListFromPrivAndPubKeys(privKIDs []string, pubKeys data.KeyList) (pubKeyList data.KeyList, err error) {
+
+	// if pubKeys are not provided, generate pubKeyList from provided private keys
+	if len(pubKeys) == 0 {
+		pubKeyList = make(data.KeyList, 0, len(privKIDs))
+
+		for _, keyID := range privKIDs {
+			pubKey := r.GetCryptoService().GetKey(keyID)
 			if pubKey == nil {
 				return nil, fmt.Errorf("unable to find key: %s", keyID)
 			}
 			pubKeyList = append(pubKeyList, pubKey)
 		}
+		return pubKeyList, nil
 	}
 
 	// Convert to certs (for root keys)
-	if pubKeyList, err = r.pubKeysToCerts(role, pubKeyList); err != nil {
-		return nil, err
+	pubKeyIDs := make(map[string]bool)
+	// list pubKeys
+	for _, k := range pubKeys {
+		kid, err := utils.CanonicalKeyID(k)
+		fmt.Println("key id : " + kid)
+		if err != nil {
+			return nil, fmt.Errorf("public key is invalid, %v", err)
+		}
+		pubKeyIDs[kid] = true
+		pubKeyList = append(pubKeyList, k)
 	}
 
+	for _, id := range privKIDs {
+		if exist := pubKeyIDs[id]; !exist {
```

Comment:
`if pubKeyIDs[id] {`

`pubKeyIDs` is a `map[string]bool` so indexing will return `true` or `false`, no need for the extra assignment.